lib/parsers/mpeg_parser.rb

```diff
@@ -0,0 +1,115 @@
+
+# MPEG Headers documentation:
+# http://dvd.sourceforge.net/dvdinfo/mpeghdrs.html#seq
+# http://www.cs.columbia.edu/~delbert/docs/Dueck%20--%20MPEG-2%20Video%20Transcoding.pdf
+# Useful tool to check the file information: https://www.metadata2go.com/
+class FormatParser::MPEGParser
+  include FormatParser::IOUtils
+
+  MPEG_VALUES = {
+    'mpg' => :mpg,
+    'mpeg' => :mpeg
+  }
+
+  ASPECT_RATIOS = {
+    1 => '1:1',
+    2 => '4:3',
+    3 => '16:9',
+    4 => '2.21:1'
+  }
+
+  FRAME_RATES = {
+    1 => '23.976',
+    2 => '24',
+    3 => '25',
+    4 => '29.97',
+    5 => '30',
+    6 => '50',
+    7 => '59.94',
+    8 => '60'
+  }
+
+  PACK_HEADER_START_CODE = '000001ba'.freeze
+  SEQUENCE_HEADER_START_CODE = 'b3'.freeze
+
+  def likely_match?(filename)
+    filename =~ /\.(mpg|mpeg)$/i
+  end
+
+  def call(io)
+    return unless matches_mpeg_header?(io)
+
+    # We are looping though the stream because there can be several sequence headers and some of them are not usefull.
+    # If we detect that the header is not usefull, then we look for the next one
+    # If we reach the EOF, then the mpg is likely to be corrupted and we return nil
+    loop do
+      search_for_next_sequence_header(io)
+      horizontal_size, vertical_size = parse_image_size(io)
+      radio_hex, rate_hex = parse_rate_information(io)
+
+      if valid_aspect_radio?(radio_hex) && valid_frame_rate?(rate_hex)
+        return file_info(horizontal_size, vertical_size, radio_hex, rate_hex)
+      end
+    end
+  rescue FormatParser::IOUtils::InvalidRead
+    nil
+  end
+
+  def file_info(width_px, height_px, radio_hex_raw, rate_hex_raw)
+    FormatParser::Video.new(
+      format: :mpg,
+      width_px: width_px,
+      height_px: height_px,
+      intrinsics: {
+        aspect_radio: ASPECT_RATIOS.fetch(radio_hex_raw),
+        frame_rate: FRAME_RATES.fetch(rate_hex_raw)
+      },
+    )
+  end
+
+  # The following 3 bytes after the sequence header code, gives us information about the px size
+  # 1.5 bytes (12 bits) for horizontal size and 1.5 bytes for vertical size
+  def parse_image_size(io)
+    image_size = to_hex(safe_read(io, 3))
+    [to_decimal(image_size[0..2]), to_decimal(image_size[3..5])]
+  end
+
+  # The following byte gives us information about the aspect ratio and frame rate
+  # 4 bits corresponds to the aspect ratio and 4 bits to the frame rate code
+  def parse_rate_information(io)
+    rate_information = to_hex(safe_read(io, 1))
+    [rate_information[0].to_i, rate_information[1].to_i]
+  end
+
+  def valid_aspect_radio?(radio)
+    ASPECT_RATIOS.include?(radio)
+  end
+
+  def valid_frame_rate?(rate)
+    FRAME_RATES.include?(rate)
+  end
+
+  ## Searches for the next SEQUENCE_HEADER_START_CODE code, which is the code for a sequence header
+  ## After this code, comes usefull information about the video
+  def search_for_next_sequence_header(io)
+    loop do
```

Comment:
Is there a reasonable upper bound to how many times we want to loop? Since we are dealing with untrusted input it might be a good idea to never do infinite loops, even if there is unbounded read protection in the IO wrapper the parser receives.

Reply:
My first idea was to get that information from the header, but I couldn't find any data that would tell me how many times I need to loop. That is why I added the `rescue FormatParser::IOUtils::InvalidRead`. In case it reaches EOF, it will return nil.

Reply:
We can put a metric here (using Measurometer) to find out "how far" we usually need to loop or how many times. Also your SEQUENCE_HEADER_START_CODE is actually a single byte. You can get better performance if you read (not `safe_read` since it is supposed to guarantee that you have received the amount of bytes you have requested) in larger chunks - say 1KB apiece. Then you can look for the sequence header start code byte inside of it and then `seek` to that place before breaking out of the loop. The reason I am suggesting this is that the IO stack in FormatParser has a limiter on how many `read` calls you perform. If it sees an unreasonable number of `read` calls it assumes the parser has been led into an infinite loop and at some point it will raise an exception. Since all you are doing is looking for a single byte (which you cannot "partially read") reading in bulk and scanning the string you have read will work nicer _and_ keep you under the read count limits.

Re. loop upper bound - if you read by 1024 bytes per iteration, you can for example say that you iterate no more than 1024 times - which will give you an upper bound but also you will know that you are going to be reading at most 1MB of data before giving up.

Reply:
Let me know what do you think about the new approach taking this into account :)